SinatraでシンプルなWebアプリを作ろう

heavy-stone opened the pull request:
Top (/memos)
<img src="https://github.com/heavy-stone/sinatra_memo/assets/44592900/700106b7-c049-4513-876a-6229779e58da" width=500>

New memo (/memos/new)
<img src="https://github.com/heavy-stone/sinatra_memo/assets/44592900/5e6d3d74-4a17-4b22-bb34-e68a7f4d339a" width=500>

Show memo (/memos/:uuid)
<img src="https://github.com/heavy-stone/sinatra_memo/assets/44592900/c97a640f-e82a-4209-b05d-6b5a4d1e1817" width=500>

Edit memo (/memos/:uuid/edit)
<img src="https://github.com/heavy-stone/sinatra_memo/assets/44592900/c7bc9956-0c54-4b67-9bdd-9d87909446bd" width=500>

heavy-stone committed:
add sinatra app with bundler

heavy-stone committed:
add rubocop-fjord

heavy-stone committed:
add erb_lint

heavy-stone committed:
add sinatra/reloader

heavy-stone committed:
add pages and json for db

heavy-stone committed:
add 404 not found

heavy-stone committed:
add minitest

heavy-stone committed:
add CGI.escapeHTML for XSS

heavy-stone committed:
add tailwindcss

heavy-stone committed:
add nl2br

heavy-stone committed:
fix erblint warnings

heavy-stone committed:
add Rakefile for test

heavy-stone committed:
update initial data

heavy-stone committed:
update README.md

JunichiIto commented on db/memos.json:
```diff
@@ -0,0 +1,17 @@
+{
```
このファイルは開発者によって内容が異なるファイルなのでコミットせずにgitignoreに追加した方がいいですね

heavy-stone replied:
下記コミットにて対応しました！
remove db/memos.json, db/memos_test.json [f085c5a](https://github.com/heavy-stone/sinatra_memo/pull/1/commits/f085c5a8b2c2c9cbb401d407fd83917044632ca6)
add .gitignore by gitignore.io [534490b](https://github.com/heavy-stone/sinatra_memo/pull/1/commits/534490b8fc9b086adaae78af90a91a124a55896a)
add db/memos.json, db/memos_test.json to .gitignore [32343e6](https://github.com/heavy-stone/sinatra_memo/pull/1/commits/32343e636f6ad255b09aeb28ac09353e9ad9eb41)

JunichiIto commented on memo.rb:
```diff
@@ -0,0 +1,89 @@
+post '/memos' do
+  @memos = read_memos_table
+  public_id = SecureRandom.uuid
+  @memos[public_id] = { public_id:, title: CGI.escapeHTML(params[:title]), content: CGI.escapeHTML(params[:content]) }
```
保存したデータをエスケープ処理すると、このアプリ以外ではデータを再利用しづらくなります。
（たとえばこのデータファイルを使ってターミナルに出力したくなった場合など）

なのでエスケープ処理は画面出力のタイミングで実施してください。

heavy-stone replied:
データは基本生データを保存しておき、DBから取り出して画面表示時にエスケープ処理をするということですね！
c12c08f にて対応しました！

JunichiIto commented on memo.rb:
```diff
@@ -0,0 +1,89 @@
+  write_memos_table(@memos)
+
+  redirect "/memos/#{public_id}"
+  erb :show
```
redirectの意味を正しく理解できていないかも？
24-25行目は何かがおかしいのでもう一度確認してみてください〜。
（patch/deleteも同様です）

heavy-stone replied:
すみません、不要なコードが残ったままとなっていました。
redirectした時点で指示したパスをGETしにいくで、その下のコードは常に実行されない認識です🙇‍♂️
30e44b1 にて対応しました！

JunichiIto commented on memo.rb:
```diff
@@ -0,0 +1,89 @@
+  memos = read_memos_table
+  return not_found if memos[public_id.to_sym].nil?
+
+  memos[public_id.to_sym]
```
`memos[public_id.to_sym]`が2回出てくるのでDRYにしましょう

heavy-stone replied:
d3df195 にて対応しました！

JunichiIto commented on Gemfile:
```diff
@@ -0,0 +1,15 @@
+gem 'sinatra'
+gem 'sinatra-contrib'
+
+group :development do
```
Railsだとgroupを指定が有効ですが、Sinatraはどうか？というのを確認してみてほしいです〜。

heavy-stone replied:
こちらについて[3/12の日報](https://bootcamp.fjord.jp/reports/91101#gemfile-rails-group-sinatra)にまとめてみましたのでご確認をお願いします！

また、SinatraとBundlerとの連携と書いた手前、連携したコードで提出したかったのですが連携したコードにすると今度は一括テスト（rake）でsinatraのmethod redefineの警告が出るようになってしまいましたので元々のコードに戻しました。
連携コードの例（memo.rb）
```ruby
require 'bundler/setup'
Bundler.require(:default, ENV['APP_ENV'] || 'development')

require 'sinatra/reloader' if development?
...省略...
```
一括テストの出力
```bash
❯ bundle exec rake
/Users/piyo/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/sinatra-4.0.0/lib/sinatra/base.rb:2077: warning: method redefined; discarding old config_file
/Users/piyo/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/sinatra-4.0.0/lib/sinatra/base.rb:2077: warning: previous definition of config_file was here
...省略...
```
上記は[調査する](https://github.com/sinatra/sinatra/issues/1319#issuecomment-1196876189)とsinatraのDelegatorの問題のようです。
解消方法が分からなかったので元々の`require 'sinatra'`に戻すと警告メッセージは表示されませんでした。

JunichiIto commented on test/constants.rb:
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+ENV['APP_ENV'] = 'test'
```
質問）この定義を無くすとテストが動かなくなったりしますか？（無くせそうな気がしたので）

heavy-stone replied:
レビュー頂いた時点で上記を削除してテストを実施しましたが、問題なく動きました！
現在は 14b4aab の対応で環境変数自体は残して活用しています。

JunichiIto commented on test/memo_test.rb:
```diff
@@ -0,0 +1,155 @@
+  end
+
+  def setup
+    File.exist?('./db/memos_test.json') && File.delete('./db/memos_test.json')
```
毎回deleteして作り直すなら`db/memos_test.json`もgitignoreに入れた方がいいですね

heavy-stone replied:
下記コミットにて対応しました！
remove db/memos.json, db/memos_test.json [f085c5a](https://github.com/heavy-stone/sinatra_memo/pull/1/commits/f085c5a8b2c2c9cbb401d407fd83917044632ca6)
add .gitignore by gitignore.io [534490b](https://github.com/heavy-stone/sinatra_memo/pull/1/commits/534490b8fc9b086adaae78af90a91a124a55896a)
add db/memos.json, db/memos_test.json to .gitignore [32343e6](https://github.com/heavy-stone/sinatra_memo/pull/1/commits/32343e636f6ad255b09aeb28ac09353e9ad9eb41)

JunichiIto replied:
```ruby
File.delete('./db/memos_test.json') if File.exist?('./db/memos_test.json')
```

という書き方の方が一般的ですねー

あと、こんなふうにしてDRYにするとか

```ruby
db_path = './db/memos_test.json'
File.delete(db_path) if File.exist?(db_path)
```

heavy-stone replied:
4813372 にて対応し、db_pathは複数ファイルで使用していたので定数化しました！

JunichiIto commented on test/memo_test.rb:
```diff
@@ -0,0 +1,155 @@
+    memos = {
+      "a90cc4d4-8a66-4fc6-8609-a02f7fe0cf96": {
+        "public_id": 'a90cc4d4-8a66-4fc6-8609-a02f7fe0cf96',
+        "title": 'test a title',
```
テストデータはなるべくリアルなデータを入れた方がテストコードが読みやすくなります。
参考 https://qiita.com/jnchito/items/2a5d3e15761fd413657a#%E3%83%86%E3%82%B9%E3%83%88%E3%82%B3%E3%83%BC%E3%83%89%E3%82%92%E6%9B%B8%E3%81%8F%E3%81%A8%E3%81%8D%E3%81%AB%E6%84%8F%E8%AD%98%E3%81%99%E3%81%B9%E3%81%8D%E3%81%93%E3%81%A8

> 実際のユースケースが想像できるテストデータを用意する

heavy-stone replied:
883f7cb e62556b にて対応しました！
参考記事ありがとうございます！
create, update, deleteのテストのbefore, afterも追加しました！

JunichiIto commented on test/memo_test.rb:
```diff
@@ -0,0 +1,155 @@
+        "content": 'test b content'
+      }
+    }
+    self.class.stub_any_instance(DB_PATH_METHOD, TEST_DB_PATH) do
```
```ruby
def db_path
  ENV['APP_ENV'] == 'test' ? './db/memos_test.json' : './db/memos.json'
end
```

みたいにすると、stubの必要がなくなってテストコードがスッキリしませんかね？

heavy-stone replied:
14b4aab にて対応しました！
stub_any_instanceの使用箇所が減り、テストコードが見やすくなりました🙌
stub_any_instanceのgemも不要になりましたので、 15ffa78 にて削除しました！

JunichiIto reviewed (commented):
コメントしました〜。

heavy-stone committed:
remove db/memos.json, db/memos_test.json

heavy-stone committed:
add .gitignore by gitignore.io

heavy-stone committed:
add db/memos.json, db/memos_test.json to .gitignore

heavy-stone committed:
add init_db_file method

heavy-stone committed:
remove stub_any_instances on tests

heavy-stone committed:
move CGI.escapeHTML to view from controller

heavy-stone committed:
remove unneeded erb after redirect

heavy-stone committed:
improve duplicate codes

heavy-stone committed:
remove minitest-stub_any_instance gem

heavy-stone committed:
add rake gem and update rake command

heavy-stone committed:
update xss test

heavy-stone committed:
improve test data to real data

heavy-stone committed:
add test group to Gemfile

heavy-stone committed:
use sinatra reloader when development

JunichiIto commented on Gemfile.lock:
```diff
@@ -127,15 +132,19 @@ GEM
+  arm64-darwin
+  x86-linux
+  x86_64-darwin
+  x86_64-linux
```
platformsがたくさん増えているのってなぜなんでしょう？

heavy-stone replied:
すみません、結論として分かりませんでした。
`bundle lock --remove-platform ***` などで消してみましたが、`bundle install` するとまた表示されてしまいます。

[Platform-specific gem ignored and Gemfile.lock unexpectedly changed by bundle install after upgrading to Ruby 3.3.0 · Issue #7506 · rubygems/rubygems]
`https://github.com/rubygems/rubygems/issues/7506`
調査したところ、どうやらRuby3.3.0の時にbundle installするとプラットフォーム固有の gem が無視され、Gemfile.lock が予期せず変更されるようです。

自分の環境もRuby3.3.0で実施していましたが、逆になぜ以前は下記の状態だったのか分かりませんでした。
```bash
PLATFORMS
  arm64-darwin-23
  ruby
```
ただ、上記Gemfile.lockをコミットした後、Bundlerの調査でBundler自体をインストールし直したり、Gemfile.lockを削除したりしたので、そのせいで発生するようになったかもしれません🙇‍♂️

ちなみに、rbenvでRuby3.2.2に変更してBundle installをすると`arm64-darwin-23` のみPLATFORMに表示されました。（rubyの文字はなし。ただその際入っていたBundlerのバージョンは2.4.18）

heavy-stone replied:
edd432d にて一応実行時のruby versionが分かるよう.ruby-versionファイルをアップしました。

JunichiIto replied:
説明ありがとうございます！
なるほど、なんか奇妙ですね。。
僕の手元でも確認してみましたが、PLATFORMSは特に増えませんでした。

ちょっと以下の操作を試してもらえますか？
（新しくローカルにコードをgit cloneして、 minitest-stub_any_instance を削除する）

```
mkdir (new dir)
cd (new dir)

git clone git@github.com:heavy-stone/sinatra_memo.git
# or
git clone https://github.com/heavy-stone/sinatra_memo.git

cd sinatra_memo

git checkout -b memo origin/memo

# https://github.com/heavy-stone/sinatra_memo/commit/15ffa782898598519665ccf3a9c1cfe14296aaee の直前に戻る
git reset --hard d3df19524fab223ff103a016c5684ac265b2bdd0

# PLATFORMSがarm64-darwin-23とrubyの2つだけであることを確認
cat Gemfile.lock

bundle install

# 何もdiffが増えてないことを確認（もしdiffが発生していたらここでストップ）
git diff

# Gemfileからgem 'minitest-stub_any_instance'の行を削除
vi Gemfile

# bundle updateではなくbundle install
# https://qiita.com/jnchito/items/823323e276d1a90ac4a1
bundle install

# diffの内容を確認
git diff
```

最後のgit diffでは僕の環境ではこんなふうになっています。
heavy-stoneさんの環境ではどうなりますか？

```diff
diff --git a/Gemfile b/Gemfile
index e4a13c6..533da6a 100644
--- a/Gemfile
+++ b/Gemfile
@@ -9,7 +9,6 @@ gem 'sinatra-contrib'
 group :development do
   gem 'erb_lint', require: false
   gem 'minitest'
-  gem 'minitest-stub_any_instance'
   gem 'rack-test'
   gem 'rubocop-fjord', require: false
 end
diff --git a/Gemfile.lock b/Gemfile.lock
index eef8049..ba6cf2f 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -50,7 +50,6 @@ GEM
       nokogiri (>= 1.12.0)
     mini_portile2 (2.8.5)
     minitest (5.22.2)
-    minitest-stub_any_instance (1.0.3)
     multi_json (1.15.0)
     mustermann (3.0.0)
       ruby2_keywords (~> 0.0.1)
@@ -133,7 +132,6 @@ PLATFORMS
 DEPENDENCIES
   erb_lint
   minitest
-  minitest-stub_any_instance
   rack-test
   rackup
   rubocop-fjord
```

heavy-stone replied:
上記の操作をしたところ、PLATFORMSは追加されず、上記と同じ差分となりました！
すみません、多分自分のローカル環境がおかしくなっているようですので、一度プロジェクトを削除してクローンして進めようと思います。
実際にご確認いただき、ありがとうございました🙇‍♂️

```diff
diff --git a/Gemfile b/Gemfile
index e4a13c6..533da6a 100644
--- a/Gemfile
+++ b/Gemfile
@@ -9,7 +9,6 @@ gem 'sinatra-contrib'
 group :development do
   gem 'erb_lint', require: false
   gem 'minitest'
-  gem 'minitest-stub_any_instance'
   gem 'rack-test'
   gem 'rubocop-fjord', require: false
 end
diff --git a/Gemfile.lock b/Gemfile.lock
index eef8049..ba6cf2f 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -50,7 +50,6 @@ GEM
       nokogiri (>= 1.12.0)
     mini_portile2 (2.8.5)
     minitest (5.22.2)
-    minitest-stub_any_instance (1.0.3)
     multi_json (1.15.0)
     mustermann (3.0.0)
       ruby2_keywords (~> 0.0.1)
@@ -133,7 +132,6 @@ PLATFORMS
 DEPENDENCIES
   erb_lint
   minitest
-  minitest-stub_any_instance
   rack-test
   rackup
   rubocop-fjord
```

また、下記で別途tmpディレクトリでbundle initからGemfile.lockの生成を確認しましたが、PLATFORMSの記載は正常になっていました。
![Monosnap Gemfile lock — tmp 2024-03-15 13-58-23](https://github.com/heavy-stone/sinatra_memo/assets/44592900/40479512-1d06-412f-982c-a3fbc1f6ea48)

heavy-stone replied:
afc382f にて対応しました！
Gemfile.lockは自動生成のため、基本的にはコマンドで操作するとのことなので、下記コマンドを投入して操作しました！
```bash
bundle lock --remove-platform aarch64-linux
bundle lock --remove-platform arm-linux
bundle lock --remove-platform arm64-darwin
bundle lock --remove-platform x86-linux
bundle lock --remove-platform x86_64-darwin
bundle lock --remove-platform x86_64-linux
bundle lock --add-platform arm64-darwin-23
bundle lock --add-platform ruby
```

JunichiIto commented on memo.rb:
```diff
@@ -0,0 +1,106 @@
+  "./db/#{db_file}"
+end
+
+def init_db_file
+  return if File.exist?(db_path)
+
+  File.open(db_path, 'w') do |file|
+    JSON.dump({}, file)
+  end
+end
```
```suggestion
def init_db_file
  return if File.exist?(db_path)

  File.open(db_path, 'w') do |file|
    JSON.dump({}, file)
  end
end
init_db_file
```

DBの初期化はアプリの起動時に一回実行すれば十分なんじゃないかなーと思います。

heavy-stone replied:
確かにそうですね、 5803963 にて対応しました！

JunichiIto commented on memo.rb:
```diff
@@ -0,0 +1,106 @@
+require 'securerandom'
+
+get '/memos' do
+  @memos = read_memos_table
```
```suggestion
  @memos = read_memos
```

https://bootcamp.fjord.jp/pages/readable-variable-name を参考に〜。

heavy-stone replied:
ハッシュの命名に悩んでいますが、`*_table`や`*_hash`や`{key}_to_{value}`などで変数名を見た時にハッシュと分かるような命名をした方がいいのかと思っていましたが名前が長くなりますし、配列と同じくデータの複数形の命名の方が分かりやすいという認識でよろしかったでしょうか？
e08b286 にて対応しました！

JunichiIto replied:
ハッシュは「ただの複数形」として命名されるケースもよくあるので、memosでも問題ないと思いますー。

JunichiIto commented on memo.rb:
```diff
@@ -0,0 +1,106 @@
+end
+
+post '/memos' do
+  @memos = read_memos_table
```
この場合、インスタンス変数に入れる必要があるのかどうかを考えてみましょう。

heavy-stone replied:
すみません、view側にデータを渡す必要がないのでローカル変数で処理して問題ないと認識しました！
a88f4db にて対応しました！

JunichiIto commented on test/memo_test.rb:
```diff
@@ -0,0 +1,166 @@
+  def test_index
+    get '/memos'
+    assert last_response.status, 200
+    assert last_response.body.include?('買い物リスト')
```
```suggestion
    assert_includes last_response.body, '買い物リスト'
```

みたいに書けるかも？（テストに失敗したときにわかりやすく表示されそう）

heavy-stone replied:
62181d9 にて対応しました！

JunichiIto commented on test/memo_test.rb:
```diff
@@ -0,0 +1,166 @@
+      }
+    }
+    assert_equal expected_before, read_memos_table
+    SecureRandom.stub(:uuid, 'c1e3e3e3-8a66-4fc6-8609-a02f7fe0cf96') do
```
```suggestion
    SecureRandom.stub(:uuid, 'new-uuid') do
```

ぐらいの方がスッキリして見やすいかも

heavy-stone replied:
fce9f0a にて対応しました！

JunichiIto commented on test/memo_test.rb:
```diff
@@ -0,0 +1,166 @@
+
+  def test_create
+    expected_before = {
+      "a90cc4d4-8a66-4fc6-8609-a02f7fe0cf96": {
+        "public_id": 'a90cc4d4-8a66-4fc6-8609-a02f7fe0cf96',
```
```suggestion
      "test-uuid-1": {
        "public_id": 'test-uuid-1',
```

ぐらいでもいいかも。テストデータなので。（リアルなuuidはコードを読むときのノイズになる）

heavy-stone replied:
e026346 にて対応しました！

JunichiIto reviewed (commented):
コメントしました〜。

heavy-stone committed:
use 3.3.0 ruby version

heavy-stone committed:
fix init_db_file to execute only when app starts

heavy-stone committed:
improve read, write method names

heavy-stone committed:
use local variables instead of instance variables

heavy-stone committed:
improve test data initialization

heavy-stone committed:
use assert_includes instead of assert

heavy-stone committed:
update uuid when test_create

heavy-stone committed:
use test uuid instead of real uuid

JunichiIto commented on memo.rb:
```diff
@@ -0,0 +1,106 @@
+
+patch '/memos/:public_id' do
+  memos = read_memos
+  memos[params[:public_id].to_sym] = { public_id: params[:public_id], title: params[:title], content: params[:content] }
```
Hash#sliceを使うともうちょっとシンプルに書けるかも？

heavy-stone replied:
5755548 にて対応しました！

JunichiIto commented on memo.rb:
```diff
@@ -0,0 +1,106 @@
+
+def read_memos
+  File.open(db_path) do |file|
+    JSON.parse(file.read, symbolize_names: true)
```
symbolize_names しない方がコードがシンプルになるのでは？と思いました

heavy-stone replied:
文字列よりシンボルの方が内部的に高速とのことなので、シンボルの方を使えるところでは使った方がいいと思っていましたが、文字列の方が分かりやすい場合は文字列のままの方がいいでしょうか？
f005513 にて対応しました！

JunichiIto replied:
good questionですね！
たしかにシンボルの方が速いというメリットはあるのですが、いつでも絶対シンボルにすべきか、というとそうではないと思います。

パフォーマンスをそこまで要求されないアプリなら、どちらでも体感できるような違いはないはずです。
文字列でもシンボルでもコード量が変わらないならシンボルを優先的に選べばいいと思いますが、今回のように文字列のままにしておけばシンプルなのに、わざわざシンボルを使おうとしてコード量が増えるのは、いわゆる「早すぎる最適化（調べてみましょう！）」に該当しそうです。

それに、キーをシンボルに揃えるなら、createの処理も`to_sym`しないとデータ型が揃わないですよね？（そしてまたコード量が増える）

```ruby
  public_id = SecureRandom.uuid.to_sym
  memos[public_id] = { public_id:, title: params[:title], content: params[:content] }
```

これは計測してみないとわかりませんが（注：この業界には「推測するな、計測せよ」という格言があります）、毎回`to_sym`を呼び出すぶんのコストが増えて、実は文字列のままの方が速かった、という可能性も無きにしも非ずです。

上記のような理由から今回は文字列のままにすることを推奨したのでした。

heavy-stone replied:
コメントありがとうございます！

> 今回のように文字列のままにしておけばシンプルなのに、わざわざシンボルを使おうとしてコード量が増える

ここの見極めができるようになるのが大切で、to_symがいるかいらないか要確認するようにしたいと思いました！
早すぎる最適化について調査してみます！

heavy-stone replied:
早すぎる最適化について調査して[日報](https://bootcamp.fjord.jp/reports/91382#-2)にまとめました🙏
きれいなコードを書くこと、無意味な最適化はしないよう心がけていきます！

JunichiIto commented on memo.rb:
```diff
@@ -0,0 +1,106 @@
+def find_memo(public_id)
+  memos = read_memos
+  memo = memos[public_id.to_sym]
+  return not_found if memo.nil?
+
+  memo
```
```suggestion
  memo or not_found
```

みたいに書くこともできそうです（ちょっとマニアックな書き方ですが）

heavy-stone replied:
こういう書き方もあるんですね！
16f03c0 にて対応しました！

JunichiIto commented on test/memo_test.rb:
```diff
@@ -0,0 +1,166 @@
+  end
+
+  def test_new
+    get '/memos/new', name: 'Frank'
```
`name: 'Frank'`って必要なんですか？

heavy-stone replied:
すみません、そもそも不要なパラメータが入っていました。
d5ca5ea にて対応しました！

JunichiIto commented on test/memo_test.rb:
```diff
@@ -0,0 +1,166 @@
+      "test-uuid-1": {
+        "public_id": 'test-uuid-1',
+        "title": '買い物リスト',
+        "content": 'トマトジュース\nティッシュペーパー'
```
```suggestion
        "content": "トマトジュース\nティッシュペーパー"
```

にしないと`\n`はただの文字列として扱われます

heavy-stone replied:
c6c38cf にて対応しました🙇‍♂️

JunichiIto commented on test/xss_test.rb:
```diff
@@ -0,0 +1,49 @@
+    memos = {
+      "test-uuid-1": {
+        "public_id": 'test-uuid-1',
+        "title": '<script>alert(\'タイトル\')</script>',
+        "content": '<script>alert(\'内容\')</script>'
```
```suggestion
        "title": '<script>alert("タイトル")</script>',
        "content": '<script>alert("内容")</script>'
```

の方がシンプルかも

heavy-stone replied:
c6c38cf にて対応しました🙇‍♂️

JunichiIto reviewed (commented):
コメントしました！

heavy-stone committed:
fix PLATFORMS on Gemfile.lock

bundle lock --remove-platform aarch64-linux
bundle lock --remove-platform arm-linux
bundle lock --remove-platform arm64-darwin
bundle lock --remove-platform x86-linux
bundle lock --remove-platform x86_64-darwin
bundle lock --remove-platform x86_64-linux
bundle lock --add-platform arm64-darwin-23
bundle lock --add-platform ruby

heavy-stone committed:
use Hash#slice

heavy-stone committed:
remove symbolize_names from JSON.parse

heavy-stone committed:
use or instead of guard clause

heavy-stone committed:
remove unneeded params

heavy-stone committed:
use " instead of '

JunichiIto commented on memo.rb:
```diff
@@ -58,16 +58,14 @@
-  return not_found if memo.nil?
-
-  memo
+  memo = memos[public_id]
+  memo or not_found
```
```suggestion
  memos[public_id] or not_found
```

でもいいですね

JunichiIto replied:
```ruby
read_memos[public_id] or not_found
```

でもいけそう

JunichiIto reviewed (approved):
1点コメントしましたが大きな問題ではないのでこれでOKです！🙆‍♂️
